### PR TITLE
Initial version

### PR DESCRIPTION
**Pull request by ddragosd**
- sync with `rclone`.  Note that `rclone` [doesn't work](https://github.com/ncw/rclone/issues/85) yet with IAM Roles and this is why we're using AWS cli. 
- includes AWS CLI for syncing files form S3

### CODE REVIEW

**Comment by clehene** [comment]
:+1:

**Comment by selfxp** [comment]
:+1:

File: config-supervisor.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"runtime"
+	"runtime/pprof"
+	"time"
+	//    "net/http"
```
Comment (adragomir) [inline config-supervisor.go]: Remove please

File: config-supervisor.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	reloadCmd    = pflag.StringP("reload-cmd", "", "echo reload-cmd not defined", "Command used to reload the gateway")
+	httpAddr     = pflag.StringP("http-addr", "", "127.0.0.1:8888", "Http Address exposing a /health-check for the sync process")
+	// when was the reload cmd executed last time
+	lastReload = time.Now()
+	// when did the last change occur
+	lastChange = time.Now()
+)
+
+func syntaxError() {
+	fmt.Fprintf(os.Stderr, `Execute a sync command and watch a folder for changes.
```
Comment (adragomir) [inline config-supervisor.go]: Move to same line. Is this after go fmt ?

File: sync/recursiveWatcher.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
+package sync
```
Comment (adragomir) [inline sync/recursiveWatcher.go:1]: the package file convention for go is lowercase ?

File: sync/recursiveWatcher.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	watcher.Folders <- folder
+}
+
+func (watcher *RecursiveWatcher) Run(debug bool) {
+	go func() {
+		for {
+			select {
+			case event := <-watcher.Events:
+				// create a file/directory
+				if event.Op&fsnotify.Create == fsnotify.Create {
```
Comment (adragomir) [inline sync/recursiveWatcher.go:65]: Separate & with spaces

File: sync/recursiveWatcher.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+func Subfolders(path string) (paths []string) {
+	filepath.Walk(path, func(newPath string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			name := info.Name()
+			// skip folders that begin with a dot
+			if shouldIgnoreFile(name) && name != "." && name != ".." {
```
Comment (adragomir) [inline sync/recursiveWatcher.go]: move comparison to shouldIgnoreFile ?

**Comment by adragomir** [comment]
@ddragosd :metal: after the comments. I'll create an issue to use sync from the go AWS SDK, I really don't like executing aws s3 sync from the command line.

**Comment by ddragosd** [comment]
@adragomir I've also started initially at integrating a library as opposed to executing a command line. I've picked [rclone](https://gowalker.org/github.com/ncw/rclone) as it comes with broader support, it's not just for AWS. 

There are 2 problems with `rclone`:
1. I couldn't integrate it as a lib - hence the idea of executing a command line
2. It doesn't work with IAM Roles :smile:  - which is a big miss for people using AWS. If the 1st point would have worked we could have fixed the IAM roles.

So at the moment I don't have a solution to support multiple clouds and integrate it as a lib, but I agree with you that it would be a nicer implementation.

File: sync/recursiveWatcher.go (hunk 3 above)
Comment (ddragosd) [inline sync/recursiveWatcher.go:1]: yes, that's what I noticed

File: sync/recursiveWatcher.go (hunk 4 above)
Comment (ddragosd) [inline sync/recursiveWatcher.go:65]: @adragomir I came to learn this is a result of running  `gofmt -e -w ./`
or `make format`

Comment (ddragosd) [inline sync/recursiveWatcher.go:65]: For Ref: https://golang.org/cmd/gofmt/

**Comment by ddragosd** [comment]
@selfxp @clehene @adragomir I've added unit tests and along with them I've enriched the `health-check` API.  This API may be exposed also through the API Gateway's `health-check` and even take action in the Gateway if the `lastSync` is not as expected ( i.e < 10s ), and return a `5xx` error, causing the container to restart. 

(updated) Sample output for the health-check API:

```
{
    status: "OK",
    lastSync: "2016-02-24T23:10:36.148512004-08:00",
    lastChangeDetected: "2016-02-24T23:07:36.036041758-08:00",
    lastReload: "2016-02-24T23:07:41.041648492-08:00"
}
```

Let me know what you think.

**Comment by selfxp** [comment]
👍

**Pull request merged**